arrow-fx/src/test/kotlin/arrow/fx/MVarTest.kt

```diff
@@ -181,48 +183,32 @@ class MVarTest : UnitSpec() {
 
         val count = 10000
         val task = mvar.just(1).flatMap { ch -> loop(count, 0, ch) }
-        task.equalUnderTheLaw(IO.just(count), EQ())
+        task.shouldBeEq(IO.just(count), EQ())
       }
 
-      "!$label - stack overflow test" {
-        // Ignored currently StackOverflows due to ListTraverse
-        val count = 10000
+      "$label - stack overflow test" {
+        val size = 20_000
 
-        fun consumer(ch: Channel<Int>, sum: Long): IO<Long> =
-          ch.take().flatMap {
-            it.fold({
-              IO.just(sum) // we are done!
-            }, { x ->
-              // next please
-              consumer(ch, sum + x)
-            })
+        fun ioMvarLoop(i: Int): IO<MVar<ForIO, Int>> =
+          mvar.just(i + 1).flatMap { it.read() }.flatMap { ii ->
+            if (ii < size) ioMvarLoop(ii)
+            else mvar.just(ii)
           }
 
-        fun exec(channel: Channel<Int>): IO<Long> {
-          val consumerTask = consumer(channel, 0L)
-          val tasks = (0 until count).map { i -> channel.put(Some(i)) }
-          val producerTask = tasks.parSequence().flatMap { channel.put(None) }
-
-          return IO.fx {
-            val f1 = !producerTask.fork()
-            val f2 = !consumerTask.fork()
-            !f1.join()
-            !f2.join()
-          }
-        }
-
-        mvar.just(Option(0)).flatMap(::exec)
-          .unsafeRunSync() shouldBe count.toLong() * (count - 1) / 2
+        mvar.just(0)
+          .flatMap { it.read() }
+          .flatMap { ioMvarLoop(it) }
+          .flatMap { it.read() }.unsafeRunSync() shouldBe size
```

Comment:
This is not the equivalent. I checked locally and there seems to be a bug in `CancelableMVar`.

I refactored `UncancelableMVar` to fix bugs and refactored it to use `IQueue` added with the `Queue` implementation. Doing the same refactor for `UncancelableMVar` might unveil the bug there too.